strings/base_coroutine_threadpool.h

```diff
@@ -381,10 +381,23 @@ namespace winrt::impl
         template <typename T>
         void await_suspend(impl::coroutine_handle<T> handle)
         {
-            set_cancellable_promise_from_handle(handle);
+            handle_type<timer_traits> new_timer;
+            new_timer.attach(check_pointer(WINRT_IMPL_CreateThreadpoolTimer(callback, this, nullptr)));
 
-            m_handle = handle;
-            create_threadpool_timer();
+            state expected = state::idle;
+            if (m_state.compare_exchange_strong(expected, state::pending, std::memory_order_release))
+            {
+                set_cancellable_promise_from_handle(handle);
```

Comment:
There's a race condition here if somebody `Cancel()` after we change the state to `pending` but before we do the `m_timer = std::move(new_timer);`. In that case, the `Cancel()` will call `fire_immediately`, which will try to fire a null `m_timer`. The `SetThreadpoolWaitEx` will fail (with "invalid parameter"), and we will bail out and do nothing.

The problem I can't solve is how to get the canceller to wait until the `m_timer` has been set before trying to cancel it.

Reply:
I think I found it. `await_resume` spins on `m_state` until it reaches a state that means that `await_suspend` is finished. We'll have to change the state machine so we can encode this without losing the cancellation flag.

Reply:
That's a valid concern. We also want to avoid assigning it if the awaiter is not idle, so I like the suggestion here.

Shouldn't the canceller be the one spinning here however, not await_resume?

Reply:
If you can find a way to do the spinning in the canceller, that would be great. But this is a synchronization issue between `await_suspend` and `await_resume`, so I'm not seeing an obvious way to shift the work to the canceller. (But I haven't thought about it too hard yet.)

Reply:
I have what I think is a solution for timers: get rid of the state, and instead rely on `IsThreadpoolTimerSet`. Store the cancellation flag as an atomic bool.

For waits, it's more complex as there is no `IsThreadpoolWaitSet` method. Right now I use a spinlock, however there is still potential for race between `set_threadpool_wait` and `fire_immediately`. My primary concern is what happens if `set_threadpool_wait` sets the wait after the cancellation by `fire_immediately` occurs: will it never fire the callback and be stuck waiting until the event is set. Other possible race combinations should be already covered by the fact that calling `SetThreadpoolWait` will cancel the previous wait.

One potential solution is a mutex and a flag that suppresses `set_threadpool_wait`.

Alternatively, if we can guarantee that a `SetThreadpoolWait` with a timeout of 0 *immediately* submits a callback (I could not find such guarantee in docs), we could simply have the callback cancel the wait, and no-op if the state is already idle (just in case).